File: architecture/presentation/src/main/java/com/pacific/presentation/Activity.java
package com.pacific.presentation;

import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProviders;
import android.content.IntentFilter;
import android.os.Bundle;
import android.support.annotation.CallSuper;
import android.support.annotation.Nullable;
import android.support.v4.content.LocalBroadcastManager;

import javax.inject.Inject;

import dagger.android.support.DaggerAppCompatActivity;
import io.reactivex.internal.functions.ObjectHelper;

public abstract class Activity extends DaggerAppCompatActivity {
    @Inject
    protected OkBroadcastReceiver okBroadcastReceiver;

    @Inject
    protected ViewModelFactory modelFactory;
    private ViewModel realViewModel;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        IntentFilter filter = new IntentFilter();
        addBroadcastAction(filter);
        LocalBroadcastManager.getInstance(this).registerReceiver(okBroadcastReceiver, filter);
        realViewModel = ViewModelProviders.of(this, modelFactory).get(modelClass());
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        okBroadcastReceiver.clearConsumer();
        LocalBroadcastManager.getInstance(this).unregisterReceiver(okBroadcastReceiver);
    }

    @CallSuper
    protected void addBroadcastAction(IntentFilter filter) {
        if (applyFinishAction()) {
            okBroadcastReceiver.addConsumer(filter, OkBroadcastReceiver.FINISH_ACTION, ((i, ii) -> finish()));
        }
    }

    @SuppressWarnings("unchecked")
    protected final <T extends ViewModel> T fetchViewModel() {
        return (T) ObjectHelper.requireNonNull(realViewModel, "realViewModel = null");
    }

    public boolean applyFinishAction() {
        return true;
    }

    protected abstract Class<? extends ViewModel> modelClass();
}

File: architecture/arch/src/main/java/com/pacific/arch/presentation/Activity.kt
package com.pacific.arch.presentation

import android.arch.lifecycle.ViewModel
import android.arch.lifecycle.ViewModelProviders
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.support.annotation.CallSuper
import android.support.v4.content.LocalBroadcastManager
import dagger.android.support.DaggerAppCompatActivity
import javax.inject.Inject

abstract class Activity : DaggerAppCompatActivity() {
    @Inject
    lateinit var okBroadcastReceiver: OkBroadcastReceiver

    @Inject
    lateinit var modelFactory: ViewModelFactory

    @Suppress("UNCHECKED_CAST")
    private val realViewModel: ViewModel by lazy {
        ViewModelProviders.of(this, modelFactory).get<ViewModel>(modelClass() as Class<ViewModel>)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val filter = IntentFilter()
        addBroadcastAction(filter)
        LocalBroadcastManager.getInstance(this).registerReceiver(okBroadcastReceiver, filter)
    }

    override fun onDestroy() {
        super.onDestroy()
        okBroadcastReceiver.clearConsumer()
        LocalBroadcastManager.getInstance(this).unregisterReceiver(okBroadcastReceiver)
    }

    @CallSuper
    protected open fun addBroadcastAction(filter: IntentFilter) {
        if (applyFinishAction()) {
            okBroadcastReceiver.addConsumer(filter,
                    ACTION_FINISH,
                    object : OkBroadcastReceiver.Consumer {
                        override fun run(context: Context, intent: Intent) {
                            finish()
                        }
                    })
        }
    }

    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> getViewModel() = realViewModel as T

    protected open fun applyFinishAction() = true

    protected abstract fun modelClass(): Class<out ViewModel>
}
